subsys/fs/littlefs_fs.c

```diff
@@ -646,30 +658,41 @@ static int littlefs_flash_init(struct fs_littlefs *fs, uintptr_t dev_id)
 	fs->backend = (void *) *fap;
 	return 0;
 }
+#endif /* CONFIG_FS_LITTLEFS_FMP_DEV */
 
 static int littlefs_init_backend(struct fs_littlefs *fs, uintptr_t dev_id, int flags)
 {
 	int ret = 0;
 
+	__ASSERT_NO_MSG((IS_ENABLED(CONFIG_FS_LITTLEFS_FMP_DEV) && !littlefs_on_blkdev(flags)) ||
+			(IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV) && littlefs_on_blkdev(flags)));
+
+#ifdef CONFIG_FS_LITTLEFS_BLK_DEV
 	if (littlefs_on_blkdev(flags)) {
 		fs->backend = (void *) dev_id;
```

Comment:
I have noticed that all #ifdefs added within source code are around `if`. So we have something like:
```
#ifdef CONFIG_FS_LITTLEFS_BLK_DEV
	if (littlefs_on_blkdev(flags)) {
```
Zephyr provides `IS_ENABLED` macro that evaluates C preprocessor identifier to 0 or 1, which could be used here:
```
if (IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV) && littlefs_on_blkdev(flags)) {
```
This makes code look nicer, as it is not so jagged by the #fidef.

Reply:
If I understand you correctly, you mean dead-code elimination (DCE) here.
Compiler parses code before the DCE optimization. It is meant that shouldn't be undefined symbols and code has to be compilable, otherwise the compilation process will fail even if we expect DCE.
If the BLK or FMP configs are disabled, a lot of functions in this file will be undefined, i.e. we need to add stub functions (I've deleted some of them according to yours previous request) or leave the current approach or enable both configs by default, but the last option contradicts the purpose of this pull-request

Reply:
OK, you are right. I have forgotten about function definitions.